### PR TITLE
Display time saved by cache at end of build

### PR DESCRIPTION
**Pull request by lptr**
Display a short summary at the end of the build about the awesomeness of the cache like this:

```text
BUILD SUCCESSFUL in 0s
17 actionable tasks: 10 executed, 7 from cache (caching saved 2m 0s)
```

Very simple demo:

[asciicast](https://asciinema.org/a/IZqUqkWEVgFvsrpXYzbyN33KL)

### CODE REVIEW

**Comment by big-guy** [comment]
@lptr this is pretty neat.  What if we started keeping origin metadata (or just build invocation id and the execution time) for all tasks so we could do this for up-to-date tasks?  That information might be interesting for build scans too.

**Comment by lptr** [comment]
That was actually what I wanted to try next as well. I'll take a look tomorrow.

**Comment by lptr** [comment]
One thing I'm unsure of is what (if anything) to display if the cache ended up wasting time instead of saving any.

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=9386726

**Comment by lptr** [comment]
BTW, build scans already keep track of origin metadata for UP-TO-DATE tasks. What we need to track this information for UP-TO-DATE is to locally keep track of how long the task took the first time we executed it. I guess if the first time was already about loading from cache, we should use the cached artifact's execution time.

That said, I'm not sure how useful it would be to print something like:

```text
Up-to-date tasks would have taken 3m 54s to execute
```

High numbers displayed in this line could be perceived as a reward for a well-maintained build. The constant reminder could reinforce the notion that you should not run your builds with `clean`, which is something we'd like to achieve. OTOH, we might be preaching to the choir, and spam users who are already aware of the advantages of incremental build.

**Comment by lptr** [comment]
I've added the time it took to look up cache misses and storing cached results, too. Now the stats is always reported if there is an impact bigger than a second (either way). The display is more terse, too.

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=9393824

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=9394104

**Comment by wolfs** [comment]
I wonder if we could have a way to make it clear to users that the saved time is cumulative task execution time. We just showed them the wall clock time for the full build, so it may be a little bit confusing to show them cumulative task execution time for the time saved by caching.
When we show how long it "took" for caching, I wonder if it is clear to users that they are this amount of time slower than actually running the task. They may interpret the time as how long it took for loading/storing to cache and not that using the cache was actually slower than executing the tasks.
Note that we also may get some results saying that using the cache is slower than executing the task if developers are on a slow internet connection and pulling from fast CI machines.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskStateInternal.java [hunk 1]
```diff
@@ -20,13 +20,16 @@
 import org.gradle.api.internal.TaskOutputCachingState;
 import org.gradle.api.tasks.TaskState;
 
+import static org.gradle.api.internal.tasks.TaskExecutionOutcome.*;
+
 public class TaskStateInternal implements TaskState {
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskStateInternal.java]: I don't like adding the logic to compute how much a task saved to `TaskStateInternal`. We also need to add some more timers to calculate durations.
All this information is already present as `BuildOperation`s. Could we listen for those and update the time saved/spent accordingly? I guess we would need to add the original duration of the task either to `ExecuteTaskBuildOperationResult` or better to `UnpackOperationResult`.
Then we could have a build operation listener which does the counting, right? All the times should already be captured as part of build operations, shouldn't they? At least we already have the times in build scans, so I guess we know them.

**Comment by lptr** [comment]
This is now implemented in part in #3669 and via the init script here: https://gist.github.com/lptr/4b57b0e9bfe9aa66f6210ad4402960c8

**Pull request closed without merging**